### DISCUSSION TITLE
ActivitySource vs DiagnosticSource

### DISCUSSION THREAD

**Discussion by aunikitin** [open-telemetry/opentelemetry-dotnet#4421] [Q&A] [answered]
Hello!
First of all, a small context:
I want to build my own metrics for Npgsql driver and I use EF Core + Npgsql.
For example, some histogram for measurement db statements from pod instance point of view. I asked guys from npgsql about this and they said that it's better to implement it by myself using their `ActivitySource` (`NpgsqlActivitySource`). ([our discussion somewhere at the end](https://github.com/npgsql/npgsql/issues/4038)) to catch traces and did whatever I want with them. After that I was inspired, and try to do it.
I copied your `ListenerHandler` infrastructure, realized my simple trace handler (without any logic, only catching an activity), but unfortunately I got nothing.
First of all, every time, when I added my handler via `TracerProviderBuilder`:
```
var instrumentation = new NpgsqlMetrics("Npgsql", options);
builder.AddSource("Npgsql");
return builder.AddInstrumentation(() => instrumentation);
```
I cought `Microsoft.EntityFramework` as `DiagnositcListener` in my `OnNext` method. After some searching I found out, that internally we have chain of listeners, and EF Core creating their own `DiagnosticSource` when you create `DbContext`. I tried to add `DiagnosticSource("Npgsql")` by my own, naively expecting that it'll help. After that I began recieving Npgsql-diagnostic listener in my `OnNext` method, but stil without triggering my custom listener.

`DiagnosticSource != ActivitySource` that is used in Npgsql driver and my question is how to coupe with `OpenTelemetry`, `ActivitySource` (from Npgsql), `DiagnosticSource` and registering infrastructure(such as `DiagnosticSourceListener`, `DiagnosticSourceSubscriber`, `ListenerHandler`) from `OpenTelemetry`?
If I should read some paper before asking, please give me a link

**Comment by utpilla** [comment, marked as answer]
>  I asked guys from npgsql about this and they said that it's better to implement it by myself using their ActivitySource (NpgsqlActivitySource).

If they have told you to use their `ActivitySource`, there are two obvious approaches to achieving what you want:

1. You could either create an [ActivityListener](https://learn.microsoft.com/en-us/dotnet/api/system.diagnostics.activitylistener?view=net-8.0) and configure it accordingly to listen to the activities created by their `ActivitySource`. With this approach, you would have to be careful with the order of registration of `ActivityListener`s (as OpenTelemetry Tracing SDK also creates its own `ActivityListener`) and ensure that you don't mess up with the sampling of activitites.

2. You could use a [custom processor](https://github.com/open-telemetry/opentelemetry-dotnet/tree/main/docs/trace/extending-the-sdk#processor) that would emit the required metrics when the activity is started or stopped. You'll have to add this custom processor to the `TracerPorvider` so that it can call the custom processors' methods on activity start and stop. Something like this:

```csharp
using var tracerProvider = Sdk.CreateTracerProviderBuilder()
    .AddSource("Npgsql")
    .AddProcessor(new NpgSqlMetricsProcessor())
    .Build();
```

```csharp
public class NpgSqlMetricsProcessor : BaseProcessor<Activity>
{
    internal Meter CustomMeter = new Meter("CustomMeter");
    internal Histogram<double> CustomMetric;
    
    public NpgSqlMetricsProcessor()
    {
        this.CustomMetric = this.CustomMeter.CreateHistogram<double>("CustomMetric");
    }

    public override void OnStart(Activity activity)
    {
    }

    public override void OnEnd(Activity activity)
    {
        if (activity.Source.Name == "Npgsql" && activity.DisplayName == "ExpectedActivityName")
        {
            this.CustomMetric.Record(activity.Duration.TotalMilliseconds, new KeyValuePair<string, object?>("Status", activity.Status));
        }
    }
}
```

**Reply by aunikitin** [reply]
Thanks for reply. I'll try this on weekends!

**Reply by aunikitin** [reply]
I tried BaseProcessor implementation. It works perfect, thanks!
